src/platform/w3dengine/client/w3dview.cpp

```diff
@@ -1754,3 +1805,475 @@ void W3DView::Move_Camera_Along_Waypoint_Path(Waypoint *way, int frames, int shu
+        }
+    }
+
+    if (m_doingMoveCameraOnWaypointPath) {
+        for (int i = 0; i < m_mcwpInfo.num_waypoints; i++) {
+            m_mcwpInfo.camera_angle[i + 1] = m_mcwpInfo.camera_angle[0];
+        }
+    }
+}
+
+void W3DView::Camera_Mod_Final_Look_Toward(Coord3D *p_loc)
+{
+    if (!m_doingRotateCamera && m_doingMoveCameraOnWaypointPath) {
+        for (int i = 2; i <= m_mcwpInfo.num_waypoints; i++) {
+            float x1 = (m_mcwpInfo.waypoints[i - 1].x + m_mcwpInfo.waypoints[i].x) / 2.0f;
+            float y1 = (m_mcwpInfo.waypoints[i - 1].y + m_mcwpInfo.waypoints[i].y) / 2.0f;
+            Coord3D c1 = m_mcwpInfo.waypoints[i];
+            Coord3D c2 = m_mcwpInfo.waypoints[i];
+            c2.x += m_mcwpInfo.waypoints[i + 1].x;
+            c2.y += m_mcwpInfo.waypoints[i + 1].y;
+            c2.x /= 2.0f;
+            c2.y /= 2.0f;
+            float x2 = (c2.x - x1) * 0.5f + x1;
+            float y2 = (c2.y - y1) * 0.5f + y1;
+            x2 += (1.0f - 0.5f) * 0.5f * (c1.x - c2.x + c1.x - x1);
+            y2 += (1.0f - 0.5f) * 0.5f * (c1.y - c2.y + c1.y - y1);
+            Vector2 v1(p_loc->x - x2, p_loc->y - y2);
+            float length = v1.Length();
+
+            if (length >= 0.1f) {
+                float angle = GameMath::Acos(v1.X / length);
+
+                if (v1.Y < 0.0f) {
+                    angle = -angle;
+                }
+
+                angle -= 1.5707964f;
+                Norm_Angle(&angle);
+                m_mcwpInfo.camera_angle[i] = angle;
+            }
+        }
+
+        if (m_mcwpInfo.total_time_milliseconds == 1) {
+            Move_Along_Waypoint_Path(1);
+            m_doingMoveCameraOnWaypointPath = true;
+            m_doingMoveCameraAlongWaypointPath = false;
+        }
+    }
+}
+
+void W3DView::Camera_Mod_Final_Move_To(Coord3D *p_loc)
+{
+    if (!m_doingRotateCamera && m_doingMoveCameraOnWaypointPath) {
+        float x = p_loc->x - m_mcwpInfo.waypoints[m_mcwpInfo.num_waypoints].x;
+        float y = p_loc->y - m_mcwpInfo.waypoints[m_mcwpInfo.num_waypoints].y;
+
+        for (int i = 2; i <= m_mcwpInfo.num_waypoints; i++) {
+            m_mcwpInfo.waypoints[i].x += x;
+            m_mcwpInfo.waypoints[i].y += y;
+        }
+    }
+}
+
+void W3DView::Camera_Mod_Look_Toward(Coord3D *p_loc)
+{
+    if (!m_doingRotateCamera && m_doingMoveCameraOnWaypointPath) {
+        int count = m_mcwpInfo.num_waypoints - 1;
+
+        if (count < 2) {
+            count = 2;
+        }
+
+        for (int i = count; i <= m_mcwpInfo.num_waypoints; i++) {
+            Coord3D c1 = m_mcwpInfo.waypoints[i - 1];
+            c1.x += m_mcwpInfo.waypoints[i].x;
+            c1.y += m_mcwpInfo.waypoints[i].y;
+            c1.x /= 2.0f;
+            c1.y /= 2.0f;
+
+            Coord3D c2 = m_mcwpInfo.waypoints[i];
+            Coord3D c3 = m_mcwpInfo.waypoints[i];
+            c3.x += m_mcwpInfo.waypoints[i + 1].x;
+            c3.y += m_mcwpInfo.waypoints[i + 1].y;
```

Comment:
I don't quite understand how this is legal. Nothing appears to write to `m_mcwpInfo.waypoints[m_mcwpInfo.num_waypoints + 1]`, so how is this a valid read?

Reply:
This is correct, that variable is written to by Setup_Waypoint_Path.